plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/split/HudiBackgroundSplitLoader.java

```diff
@@ -51,44 +54,47 @@ public HudiBackgroundSplitLoader(
             AsyncQueue<ConnectorSplit> asyncQueue,
             Executor splitGeneratorExecutor,
             HudiSplitWeightProvider hudiSplitWeightProvider,
-            List<String> partitions)
+            List<String> partitions,
+            Consumer<Throwable> errorListener)
     {
         this.hudiDirectoryLister = requireNonNull(hudiDirectoryLister, "hudiDirectoryLister is null");
         this.asyncQueue = requireNonNull(asyncQueue, "asyncQueue is null");
         this.splitGeneratorExecutor = requireNonNull(splitGeneratorExecutor, "splitGeneratorExecutorService is null");
         this.splitGeneratorNumThreads = getSplitGeneratorParallelism(session);
         this.hudiSplitFactory = new HudiSplitFactory(tableHandle, hudiSplitWeightProvider);
         this.partitions = requireNonNull(partitions, "partitions is null");
+        this.errorListener = requireNonNull(errorListener, "errorListener is null");
     }
 
     @Override
     public void run()
     {
         Deque<String> partitionQueue = new ConcurrentLinkedDeque<>(partitions);
         List<HudiPartitionInfoLoader> splitGeneratorList = new ArrayList<>();
-        List<Future> splitGeneratorFutures = new ArrayList<>();
+        List<ListenableFuture<Void>> splitGeneratorFutures = new ArrayList<>();
 
         // Start a number of partition split generators to generate the splits in parallel
         for (int i = 0; i < splitGeneratorNumThreads; i++) {
             HudiPartitionInfoLoader generator = new HudiPartitionInfoLoader(hudiDirectoryLister, hudiSplitFactory, asyncQueue, partitionQueue);
             splitGeneratorList.add(generator);
-            splitGeneratorFutures.add(Futures.submit(generator, splitGeneratorExecutor));
+            ListenableFuture<Void> future = Futures.submit(generator, splitGeneratorExecutor);
+            addExceptionCallback(future, throwable -> errorListener.accept(throwable));
+            splitGeneratorFutures.add(future);
         }
 
         for (HudiPartitionInfoLoader generator : splitGeneratorList) {
             // Let the split generator stop once the partition queue is empty
             generator.stopRunning();
         }
 
-        // Wait for all split generators to finish
-        for (Future future : splitGeneratorFutures) {
-            try {
-                future.get();
-            }
-            catch (InterruptedException | ExecutionException e) {
-                throw new TrinoException(HUDI_CANNOT_OPEN_SPLIT, "Error generating Hudi split", e);
-            }
+        try {
+            // Wait for all split generators to finish
+            Futures.whenAllComplete(splitGeneratorFutures)
+                    .run(asyncQueue::finish, directExecutor())
+                    .get();
+        }
+        catch (Exception e) {
```

Comment:
We shouldn't swallow `InterruptedException` here (this is a bug in the existing code, but the change to `catch (Exception)` makes it harder to see:
```java
catch (InterruptedException | ExecutionException e) {
    if (e instanceof InterruptedException) {
        Thread.currentThread().interrupt();
    }
    throw new TrinoException(HUDI_CANNOT_OPEN_SPLIT, "Error generating Hudi split", e);
}

Reply:
Thanks for your explanation.